include/deal.II/base/tensor.h

```diff
@@ -1313,7 +1313,27 @@ template <typename ArrayLike, std::size_t... indices>
 constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
 Tensor<rank_, dim, Number>::Tensor(const ArrayLike &initializer,
                                    std::index_sequence<indices...>)
-  : values{{value_type(initializer[indices])...}}
+  // Extract from the 'initializer' a sequence of elements via template
+  // pack evaluation. This could be as easy as
+  //     values{{ (initializer[indices])... }}
+  // but of course in practice it is not. The challenge is that if rank>1,
+  // we want to pass the elements initializer[indices] down to the next
+  // lower rank tensor for evaluation unchanged. But at the rank==1 level,
+  // we need to convert to the scalar type 'Number'. This would all be
+  // relatively straightforward if we could rely on automatic type
+  // conversion, but for some autodifferentiation types, the conversion
+  // from the AD to double (i.e., the extraction of a scalar value) is
+  // not implicit, and we need to call internal::NumberType<Number>::value() --
+  // but as mentioned, we can only do that for rank==1.
+  //
+  // We can achieve all of this by dispatching to a lambda function within
+  // which we can use a 'if constexpr'.
+  : values{{([&initializer]() -> value_type {
+    if constexpr (rank_ == 1)
+      return internal::NumberType<Number>::value(initializer[indices]);
+    else
+      return value_type(initializer[indices]);
```

Comment:
For reference, if anyone is curious: The version that failed so miserably in the beginning did not explicitly provide the return type for the lambda function, and did not do the explicit cast to `value_type` in the last line. Because `values` is an array of elements of type `value_type`, you'd think that the compiler can do the conversion *when assigning to `values`* itself, rather than already getting an array of `value_type` objects. 

But that is not true due to the bizarre decay rules of arrays. Notably, if `ArrayType` is `double[2][2]`, then what we *want* to return is an array `double[2]` (or perhaps a reference to it), but when the compiler sees `initializer[indices]` and is asked to return that, it instead returns ` double *`, and that can not be used to initialize a `Tensor<1,dim>`.

The solution is to spell out the conversion explicitly here, so that the end result is an initializer list for an array of elements of type `value_type`. This allows the compiler to do in-place object creation for the array, without having to perform additional conversions.